### PR TITLE
feat: [PL-38632]: Adding support for Service Discovery

### PR DESCRIPTION
**Pull request by S-ayanide**
### Summary

This feature implements Service Discovery, Discovery Agent creation and Network Map implementation under project settings

#### Screenshots

Screenshot 2023-06-15 at 6 08 55 PM
Screenshot 2023-06-15 at 6 09 04 PM
Screenshot 2023-06-15 at 6 09 41 PM
Screenshot 2023-06-15 at 6 09 46 PM
Screenshot 2023-06-15 at 6 09 52 PM
Screenshot 2023-06-15 at 6 10 04 PM

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

File: src/modules/73-discovery/images/related-networks.svg
Comment (abhinavrastogi-harness) [inline src/modules/73-discovery/images/related-networks.svg:1]: icons should be in uicore/icons

File: src/modules/73-discovery/images/service-settings.svg
Comment (abhinavrastogi-harness) [inline src/modules/73-discovery/images/service-settings.svg:1]: i think we already have these icons in uicore

File: src/modules/73-discovery/pages/discovery-details/views/network-map/EmptyStateNetworkMap.tsx [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+          </div>
+          <Icon name="arrow-right" />
+          <div style={{ display: 'flex', alignItems: 'center' }}>
+            <img src={Relations} />
+            <Text margin={{ left: 'xsmall' }}>{getString('discovery.tabs.configureRelations')}</Text>
+          </div>
+        </Layout.Horizontal>
+        <Button
+          width={'220px'}
+          text="Create New Network Map"
```
Comment (abhinavrastogi-harness) [inline src/modules/73-discovery/pages/discovery-details/views/network-map/EmptyStateNetworkMap.tsx:49]: externalise all strings

File: src/modules/73-discovery/pages/discovery-details/views/network-map/NetworkMapTable.tsx [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+  }
+
+  return (
+    <Container>
+      {networkMapList?.items?.length !== 0 ? (
+        <>
+          <Page.SubHeader>
+            <Layout.Horizontal width="100%" flex={{ justifyContent: 'space-between' }}>
+              <Button
+                text="New Network Map"
```
Comment (abhinavrastogi-harness) [inline src/modules/73-discovery/pages/discovery-details/views/network-map/NetworkMapTable.tsx:140]: externalise strings

File: src/modules/73-discovery/pages/discovery-details/views/network-map/NetworkMapTable.tsx [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+                throttle={500}
+                defaultValue={search}
+                onChange={value => setSearch(value)}
+              />
+            </Layout.Horizontal>
+          </Page.SubHeader>
+          {networkMapListLoading ? (
+            <Page.Spinner />
+          ) : (
+            <TableV2<DatabaseNetworkMapCollection>
```
Comment (abhinavrastogi-harness) [inline src/modules/73-discovery/pages/discovery-details/views/network-map/NetworkMapTable.tsx:167]: all list pages should use the <ListHeader /> component with standard sorting capabilities

File: src/modules/73-discovery/pages/discovery-details/views/network-map/NetworkMapTable.tsx [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+              className={css.tableBody}
+              sortable={true}
+              columns={[
+                {
+                  Header: 'NETWORK MAPS',
```
Comment (abhinavrastogi-harness) [inline src/modules/73-discovery/pages/discovery-details/views/network-map/NetworkMapTable.tsx:172]: externalise all strings

File: src/modules/73-discovery/pages/discovery-details/views/settings/ListItems.tsx [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import { Color, FontVariation, StyledProps } from '@harness/design-system'
```
Comment (abhinavrastogi-harness) [inline src/modules/73-discovery/pages/discovery-details/views/settings/ListItems.tsx:1]: this and few other components seem duplicated from `ServiceDetails` . 
please re-use these compomnents.

File: src/modules/73-discovery/pages/discovery-details/views/settings/Settings.module.scss
Comment (abhinavrastogi-harness) [inline src/modules/73-discovery/pages/discovery-details/views/settings/Settings.module.scss:1]: please remove empty files

File: src/modules/73-discovery/pages/home/DiscoveryPage.tsx [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+                      alwaysExpanded
+                      throttle={500}
+                      defaultValue={search}
+                      onChange={value => setSearch(value)}
+                      placeholder={getString('discovery.homepage.searchDiscoveryAgent')}
+                    />
+                  </Container>
+                </Layout.Horizontal>
+              </Page.SubHeader>
+              <Page.Body className={css.discoveryAgentTable}>
```
Comment (abhinavrastogi-harness) [inline src/modules/73-discovery/pages/home/DiscoveryPage.tsx:114]: all list pages should use the <ListHeader /> component with standard sorting capabilities

File: src/modules/73-discovery/pages/home/views/create-discovery-agent/CreateDAgent.tsx [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+                serviceAccount: 'cluster-admin-1'
+              }
+            }
+          })
+            .then(() => {
+              setDrawerOpen(false)
+            })
+            .catch(e => showError(e))
+        } else {
+          showError('Invalid configuration, either required fields were not filled or validation has failed')
```
Comment (abhinavrastogi-harness) [inline src/modules/73-discovery/pages/home/views/create-discovery-agent/CreateDAgent.tsx:92]: externalise

File: src/modules/73-discovery/pages/home/views/create-discovery-agent/CreateDAgent.tsx [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+        }
+      })
+    }
+  }
+
+  return (
+    <>
+      <Layout.Horizontal
+        width="100%"
+        height="7vh"
```
Comment (abhinavrastogi-harness) [inline src/modules/73-discovery/pages/home/views/create-discovery-agent/CreateDAgent.tsx:102]: suspiciously specific. why is this using relative values? cant we hardcode the height in px?

File: src/modules/73-discovery/pages/home/views/create-discovery-agent/CreateDAgent.tsx [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+            initialValues={inputValues}
+            validationSchema={Yup.object().shape({
+              discoveryAgentName: Yup.string()
+                .trim()
+                .matches(/^[^-].*$/, 'Discovery Agent name can not start with -')
+                .matches(/^.*[^-]$/, 'Discovery Agent name can not end with -')
+                .max(50, 'Discovery Agent name can have a max length of 50 characters')
+                .required('Discovery Agent name is required!'),
+              connectorRef: Yup.string().trim().required('Please select a Connector!'),
+              discoveryNamespace: Yup.string().trim().required('Please select a discovery namespace')
```
Comment (abhinavrastogi-harness) [inline src/modules/73-discovery/pages/home/views/create-discovery-agent/CreateDAgent.tsx:138]: externalise strings

File: src/modules/73-discovery/pages/home/views/empty-state/EmptyStateDiscoveryAgent.tsx [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+      <Text width={600} style={{ textAlign: 'center' }}>
+        {getString('discovery.homepage.discoveryAgentDesc')}
+      </Text>
+      <Button
+        onClick={() => setDrawerOpen(true)}
+        text={getString('discovery.homepage.newDiscoveryAgentBtn')}
+        icon="plus"
+        variation={ButtonVariation.PRIMARY}
+      />
+      <Link to={'#'} target="_blank">
```
Comment (abhinavrastogi-harness) [inline src/modules/73-discovery/pages/home/views/empty-state/EmptyStateDiscoveryAgent.tsx:34]: broken link?

File: src/modules/73-discovery/pages/network-map-studio/NetworkMapStudio.tsx [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+  const { isOpen, open, close } = useToggleOpen()
+  const history = useHistory()
+  const createNetworkMapLabel = getString('discovery.createNetworkMap')
+  const networkMapFormRef = React.useRef<FormikProps<FormValues>>()
+
+  const [title, setTitle] = useState<string>('Untitled Network Map')
+
+  useDocumentTitle(createNetworkMapLabel)
+
+  const handleTabChange = (tabID: StudioTabs): void => {
```
Comment (abhinavrastogi-harness) [inline src/modules/73-discovery/pages/network-map-studio/NetworkMapStudio.tsx:63]: can we check if existing TabNavigation component can be reused?
https://uicore.harness.io/?path=/story/components-tabnavigation--basic

File: src/modules/73-discovery/pages/network-map-studio/views/select-service/SelectService.tsx [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+              <Layout.Vertical spacing={'medium'} style={{ alignItems: 'center' }}>
+                <Icon name="steps-spinner" size={32} color={Color.GREY_600} />
+                <Text font={{ size: 'medium', align: 'center' }} color={Color.GREY_600}>
+                  {getString('loading')}
+                </Text>
+              </Layout.Vertical>
+            </Container>
+          ) : (
+            <Container width="95%" height="75vh" style={{ margin: 'auto' }}>
+              <TableV2<DatabaseServiceCollection>
```
Comment (abhinavrastogi-harness) [inline src/modules/73-discovery/pages/network-map-studio/views/select-service/SelectService.tsx:245]: all list pages should use the standard `<ListHeader />` component with standard sorting

File: src/services/servicediscovery/index.tsx
Comment (abhinavrastogi-harness) [inline src/services/servicediscovery/index.tsx:1]: do we really need to ship all 12k lines of the service for all customers? 😅 
why have we given `*` for allowPaths?

File: src/modules/RouteDestinations.tsx [hunk 13]
```diff
@@ -14,6 +14,7 @@ import commonRoutes from '@common/RouteDestinations'
 import { useFeatureFlags } from '@common/hooks/useFeatureFlag'
 import AuthSettingsRoutes from '@auth-settings/RouteDestinations'
 import secretsRoutes from '@secrets/RouteDestinations'
+import discoveryRoutes from '@discovery/RouteDestinations'
```
Comment (abhinavrastogi-harness) [inline src/modules/RouteDestinations.tsx:17]: can we look into a dynamic import for this module? we should avoid loading all this code for 99% of the users who wont be using this right now.

**Review by abhinavrastogi-harness — CHANGES_REQUESTED** [review]

File: configs/restful-react.config.js [hunk 14]
```diff
@@ -24,6 +24,16 @@ module.exports = {
       base: `{getConfig("api")}`
     }
   },
+  'servicediscovery': {
```
[the remark this answers is not in the document]
Comment (arkajyotiMukherjee) [reply configs/restful-react.config.js:27]: We are using swagger 2.0. Do we need to create the spec in Open API v3?

File: src/services/servicediscovery/index.tsx
Comment (arkajyotiMukherjee) [reply src/services/servicediscovery/index.tsx:1]: we will change it to only required APIs

File: src/modules/73-discovery/components/NumberedList/NumberedList.tsx [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+import type { StyledProps } from '@harness/design-system'
+import css from './NumberedList.module.scss'
+
+interface NumberedListProps extends StyledProps {
+  index: number
+  showLine?: boolean
+  content: React.ReactNode
+}
+
+const NumberedList: React.FC<NumberedListProps> = props => {
```
[the remark this answers is not in the document]
Comment (S-ayanide) [reply src/modules/73-discovery/components/NumberedList/NumberedList.tsx:19]: This component is used for the design below actually, its not the regular <li> item, the numbers along with its content also takes a showLine props which determines if the left vertical lines would render or not.

Screenshot 2023-06-20 at 1 21 08 PM

File: src/modules/73-discovery/pages/discovery-details/views/discovered-resources/DiscoveredServices.tsx [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+
+export interface ConnectionMap {
+  [sourceID: string]: ApiCustomServiceConnection[]
+}
+
+export interface K8SCustomService extends DatabaseK8SCustomServiceCollection {
+  relatedServices?: ApiCustomServiceConnection[]
+}
+
+const DiscoveredServices: React.FC = () => {
```
[the remark this answers is not in the document]
Comment (amityt) [reply src/modules/73-discovery/pages/discovery-details/views/discovered-resources/DiscoveredServices.tsx:51]: Hi @abhinavrastogi-harness , currently the backend apis don't have the sorting functionality. 😅  The API dev is still under progress, once it is available, will add the <ListHeader />

File: src/modules/73-discovery/pages/network-map-studio/NetworkMapStudio.tsx (hunk 11 above)
Comment (S-ayanide) [reply src/modules/73-discovery/pages/network-map-studio/NetworkMapStudio.tsx:63]: The functionality is same with both the approaches, but the UX design specifically required a different UI, like the one below, which will not be very straight forward with TabNavigation component

Screenshot 2023-06-23 at 1 30 17 PM

File: src/modules/73-discovery/pages/network-map-studio/views/select-service/SelectService.module.scss [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+  background-size: $dot-space $dot-space !important;
+
+  .image {
+    margin: 50px auto;
+  }
+}
+
+.tableBody {
+  overflow-y: auto;
+  div[class*='TableV2--row TableV2--card'] {
```
[the remark this answers is not in the document]
Comment (S-ayanide) [reply src/modules/73-discovery/pages/network-map-studio/views/select-service/SelectService.module.scss:51]: The table design for SelectService is specifically different, the in-build styling wouldn't look the same without some extra modification

Screenshot 2023-06-23 at 1 31 07 PM

File: configs/restful-react.config.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (abhinavrastogi-harness) [reply configs/restful-react.config.js:27]: yes. we should be using open api v3 for all new apis

**Comment by S-ayanide** [comment]
retrigger titlecheck

**Comment by AnshuH** [comment]
retrigger titlecheck

File: src/modules/73-discovery/pages/home/views/create-discovery-agent/CreateDAgent.tsx [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+      dAgentFormRef.current.validateForm().then(errors => {
+        if (Object.keys(errors).length === 0) {
+          infraMutate({
+            k8sConnectorID: dAgentFormRef.current?.values.connectorRef,
+            name: dAgentFormRef.current?.values.discoveryAgentName,
+            identity: dAgentFormRef.current?.values.identifier,
+            config: {
+              kubernetes: {
+                namespace: dAgentFormRef.current?.values.discoveryNamespace,
+                imageRegistry: 'index.docker.io/shovan1995',
```
Comment (SahilKr24) [inline src/modules/73-discovery/pages/home/views/create-discovery-agent/CreateDAgent.tsx:82]: are we using this for stress also

File: src/modules/73-discovery/pages/network-map-studio/NetworkMapStudio.tsx [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+const NetworkMapStudio: React.FC = () => {
+  const { dAgentId } = useParams<DiscoveryPathProps & ModulePathParams>()
+  const { accountId, orgIdentifier, projectIdentifier } = useParams<ProjectPathProps & ModulePathParams>()
+  const { getString } = useStrings()
+  const { isOpen, open, close } = useToggleOpen()
+  const history = useHistory()
+  const createNetworkMapLabel = getString('discovery.createNetworkMap')
+  const networkMapFormRef = React.useRef<FormikProps<FormValues>>()
+
+  const [title, setTitle] = useState<string>('Untitled Network Map')
```
Comment (SahilKr24) [inline src/modules/73-discovery/pages/network-map-studio/NetworkMapStudio.tsx:59]: external strings

File: src/modules/73-discovery/pages/network-map-studio/NetworkMapStudio.tsx [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+                label: dAgentId
+              }
+            ]}
+          />
+        }
+        title={
+          <React.Fragment>
+            <String tagName="div" className={css.networkMapTitle} stringID="common.networkMap" />
+            <Layout.Horizontal spacing="small" flex={{ alignItems: 'center', justifyContent: 'flex-start' }}>
+              <Text color={title === 'Untitled Network Map' ? Color.GREY_200 : Color.GREY_900} font="medium">
```
Comment (SahilKr24) [inline src/modules/73-discovery/pages/network-map-studio/NetworkMapStudio.tsx:95]: external strings

File: src/modules/73-discovery/pages/network-map-studio/views/select-service/SelectService.tsx [hunk 21]
```diff
@@ -0,0 +1,286 @@
...
+          kind: service.kind
+        },
+        port: service.spec && service.spec.ports && service.spec.ports[0].port?.toString(),
+        to: {
+          id: nextService.id,
+          kind: nextService.kind
+        }
+      })
+    }
+    for (let index = 0; index < selectedServices.length - 1; index++) {
```
Comment (SahilKr24) [inline src/modules/73-discovery/pages/network-map-studio/views/select-service/SelectService.tsx:101]: can't we use map for this?

File: src/modules/73-discovery/pages/network-map-studio/views/select-service/SelectService.tsx [hunk 22]
```diff
@@ -0,0 +1,286 @@
...
+                    projectIdentifier,
+                    dAgentId: dAgentId
+                  })
+                )
+              }
+            />
+            <Button
+              type="submit"
+              variation={ButtonVariation.PRIMARY}
+              text={'Create Network Map'}
```
Comment (SahilKr24) [inline src/modules/73-discovery/pages/network-map-studio/views/select-service/SelectService.tsx:273]: external strings

File: src/modules/73-discovery/pages/network-map-studio/views/select-service/SelectService.tsx [hunk 23]
```diff
@@ -0,0 +1,286 @@
...
+              <TableV2<DatabaseServiceCollection>
+                className={css.tableBody}
+                columns={columns}
+                data={discoveredServices?.items ?? []}
+                pagination={paginationProps}
+              />
+            </Container>
+          )}
+        </Container>
+        <Container className={css.bottomNav} padding={'medium'} height="8vh">
```
Comment (SahilKr24) [inline src/modules/73-discovery/pages/network-map-studio/views/select-service/SelectService.tsx:254]: fixed height?

**Review by SahilKr24 — COMMENTED** [review]
Some of these could be important, please a took before merging, Thanks.

**Comment by arkajyotiMukherjee** [comment]
retrigger jest

**Review by abhinavrastogi-harness — APPROVED** [review]

**Pull request merged**